lib/python/qmk/cli/c2json.py

```diff
@@ -0,0 +1,324 @@
+    """
+    Turns a list of modifiers into parenthesized form
+
+    Parameters:
+    modifiers -- List[str] of modifiers
+    keycode -- string keycode to apply them to
+
+    Example:
+    >>> make_modifier_keycode(['LALT', 'LSFT', 'LCTL'], 'KC_A')
+    'LCTL(LSFT(LALT(KC_A)))'
+    """
+    result = keycode
+    for modifier in modifiers:
+        result = '{}({})'.format(modifier, result)
+    return result
+
+
+def extract_row(row_ast, log):
+    """
+    Extracts a row as a List[Key] of keycode strings
+
+    Transforms an InitList of keycode ASTs into a list of string QMK keycodes
+    """
+    # each layer is a
+    # InitList of rows ->
+    #   InitList of IDs or more complex objects
+    row = []
+    for expr in row_ast.exprs:
+        if type(expr) == ID:
+            row.append(expr.name)
+        elif type(expr) == Constant and expr.type == 'int':
+            if expr.value != '0x01':
+                # KC_TRNS is represented by 0x01
+                log.warn('Discarding an unrecognized numeric keycode: %s', expr.value)
+            row.append('KC_TRNS')
+        elif type(expr) == BinaryOp:
+            # it's probably a modifier set
+            modifiers, keycode = or_ast_to_modifiers(expr, log=log)
+            row.append(make_modifier_keycode(modifiers, keycode))
+        else:
+            log.warn('Not implemented: emitting KC_TRNS for %s', expr)
+            row.append('KC_TRNS')
+            continue
+    return row
+
+
+def extract_layers(keymaps, log):
+    """
+    Takes an AST representation of the keymaps array and pulls the layers out
+    as Layer objects
+
+    Returns:
+    List[Layer] representing the layers in the AST
+    """
+    layers = []
+    for layer in keymaps.init.exprs:
+        name = layer.name[0].value
+        rows_ast = layer.expr.exprs
+        rows = []
+        for row_ast in rows_ast:
+            rows.append(extract_row(row_ast, log=log))
+        layers.append(Layer(name=name, keys=rows))
+    return layers
+
+
+def find_keyboard_name_from_keymap_c(keymap_c):
+    """
+    Takes the path to a keymap.c and divines the keyboard name.
+
+    Parameters:
+    keymap_c -- Path object representing the keymap.c file
+
+    >>> find_keyboard_name_from_keymap_c(Path('keyboards/clueboard/66/keymaps/default/keymap.c'))
+    'clueboard/66'
+    """
+    parts = keymap_c.parts
+    return '/'.join(parts[(parts.index('keyboards') + 1) : parts.index('keymaps')])
+
+
+def make_keymap_json(keymap_c, layer_list):
+    """
+    Converts a list of our internal Layer objects to a JSON dictionary
+    suitable for the configurator.
+
+    Parameters:
+    keymap_c -- Path representing the location of the keymap.c
+    layer_list -- List[Layer]
+    """
+    flatten_list = lambda lst: list(itertools.chain.from_iterable(lst))
+
+    obj = {
+        'keyboard': find_keyboard_name_from_keymap_c(keymap_c),
+        'keymap': keymap_c.parent.name,
+        'layout': 'KEYMAP',  # I don't know what this means
```

Comment:
This is the `LAYOUT` macro for the keyboard. It defines the relationship between the physical layouts of keys and their position on the scanning matrix. Many keyboards support multiple layouts macros. In my example below the layout macro is `LAYOUT_all`:

```
const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
    [0] = LAYOUT_all(KC_A),
    [1] = LAYOUT_all(KC_B)
}
```

Note: `KEYMAP` is a deprecated name we try not to use anymore.